central/delegatedregistryconfig/service/service.go

```diff
@@ -0,0 +1,227 @@
+func (s *serviceImpl) RegisterServiceServer(grpcServer *grpc.Server) {
+	v1.RegisterDelegatedRegistryConfigServiceServer(grpcServer, s)
+}
+
+// RegisterServiceHandler registers this service with the given gRPC Gateway endpoint.
+func (s *serviceImpl) RegisterServiceHandler(ctx context.Context, mux *runtime.ServeMux, conn *grpc.ClientConn) error {
+	return v1.RegisterDelegatedRegistryConfigServiceHandler(ctx, mux, conn)
+}
+
+// AuthFuncOverride specifies the auth criteria for this API.
+func (s *serviceImpl) AuthFuncOverride(ctx context.Context, fullMethodName string) (context.Context, error) {
+	return ctx, authorizer.Authorized(ctx, fullMethodName)
+}
+
+// GetConfig returns Central's delegated registry config
+func (s *serviceImpl) GetConfig(ctx context.Context, _ *v1.Empty) (*storage.DelegatedRegistryConfig, error) {
+	if s.dataStore == nil {
+		return nil, status.Error(codes.Unimplemented, "datastore not initialized, is postgres enabled?")
+	}
+
+	config, err := s.dataStore.GetConfig(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "retrieving config")
+	}
+
+	if config == nil {
+		return &storage.DelegatedRegistryConfig{}, nil
+	}
+
+	return config, nil
+}
+
+// GetClusters returns the list of clusters (id + name) that are eligible for delegating scanning
+// requests (ie: only clusters with scanners that understand the delegated registry config)
+func (s *serviceImpl) GetClusters(ctx context.Context, _ *v1.Empty) (*v1.DelegatedRegistryClustersResponse, error) {
+	if s.dataStore == nil {
+		return nil, status.Error(codes.Unimplemented, "datastore not initialized, is postgres enabled?")
+	}
+
+	clusters, err := s.getClusters(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "retrieving clusters")
+	}
+
+	if len(clusters) == 0 {
+		return nil, status.Error(codes.NotFound, "no valid clusters found")
+	}
+
+	return &v1.DelegatedRegistryClustersResponse{
+		Clusters: clusters,
+	}, nil
+}
+
+// PutConfig updates Central's delegated registry config
+func (s *serviceImpl) PutConfig(ctx context.Context, config *storage.DelegatedRegistryConfig) (*storage.DelegatedRegistryConfig, error) {
+	if s.dataStore == nil {
+		return nil, status.Error(codes.Unimplemented, "datastore not initialized, is postgres enabled?")
+	}
+
+	if err := s.validate(ctx, config); err != nil {
+		return nil, errors.Wrap(errox.InvalidArgs, err.Error())
+	}
+
+	if err := s.dataStore.UpsertConfig(ctx, config); err != nil {
+		return nil, errors.Wrap(err, "upserting config")
+	}
+
+	return config, nil
+}
+
+func (s *serviceImpl) validate(ctx context.Context, config *storage.DelegatedRegistryConfig) error {
+	if config == nil {
+		// this block not reachable via GRPC-gateway invocations
+		return errors.New("config missing")
+	}
+
+	if config.EnabledFor == storage.DelegatedRegistryConfig_NONE {
+		// ignore rest of config
+		return nil
+	}
+
+	errorList := errorhelpers.NewErrorList("Validation")
```

Comment:
There is a new type of error joining which is in go 1.20: https://go.dev/blog/go1.20 where you can join errors so that errors.Is and errors.As still works

Reply:
Nice! I took a look - in this particular workflow I'm currently not inspecting the type of errors - is there a scenario I'm missing that will benefit from `errors.Join`?

Reply:
namely the one in which I want to deprecate NewErrorList :)

Reply:
ahh, didn't realize `errorList` was on the way out - ok I'll remove `errorList` in favor of a `[]error` slice and `errors.Join`

Reply:
take a look at the latest commit, that what you had in mind?